### DISCUSSION TITLE
doc example authentication

### DISCUSSION THREAD

**Discussion by tamis-laan** [TanStack/query#3253] [Ideas]
I'm trying to learn react query and the first thing I'm hitting is authentication where the user needs to register, activate account, login, logout and deregister. At the same time there is a dependency between the user state and being logged in or not. 

I feel this case is not very straight forward and has elements like token persistence and token expiring plus refresh tokens that make it less obvious as how to deal with this proper using react query.

Given the frequent occurrence of authentication it would be great if an example or best practice would be included inside the examples section of the docs.

**Comment by jmsims2** [comment]
Generally speaking, all of that would be happening in your query fn. For example, in your login query function, you could put something in `onSuccess` to put the cookie into cookie storage. For authenticating, you could have an interceptor (if using axios) or something like that, that adds all the correct headers and what not. We have a little connector function that all of our query functions use that handles adding headers and logging out if there is a 401/403 response, etc.

In other words, react query is not opinionated about how you handle this.

**Reply by tamis-laan** [reply]
It's great that react query is not an opinionated framework and I really like react query :-) 

All I'm saying is that the first step in moving a code base to react query is implementing authentication. And because authentication deals with a few topics like persisting tokens and expiring tokens it deviates from the general use case.

I notice that with me (and I think for others too) it creates a barrier of entry. For me it would have helped allot if there was an authentication example in the docs so I could get a general idea as how to implement this.

Of course given that react query is not opinionated I would definitely do things differently more suitable to my setup. But that first example would provide an entry point and clear up some questions that arise naturally when implementing authentication with react query.

**Comment by bleriotnoguia** [comment]
Maybe this link can help https://codevoweb.com/react-query-context-api-axios-interceptors-jwt-auth/
React Query, and Axios Interceptors JWT Authentication
